druntime/src/core/stdc/stdatomic.d

```diff
@@ -0,0 +1,1124 @@
+                set = get | arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_acquire, memory_order.memory_order_acquire)(cast(A*)obj, &get, cast(A)set));
+            break;
+
+        case memory_order.memory_order_release:
+            get = atomicLoad!(memory_order.memory_order_relaxed)(cast(A*)obj);
+            do
+            {
+                set = get | arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_release, memory_order.memory_order_release)(cast(A*)obj, &get, cast(A)set));
+            break;
+
+        case memory_order.memory_order_acq_rel:
+            get = atomicLoad!(memory_order.memory_order_acq_rel)(cast(A*)obj);
+            do
+            {
+                set = get | arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_acq_rel, memory_order.memory_order_acq_rel)(cast(A*)obj, &get, cast(A)set));
+            break;
+
+        case memory_order.memory_order_seq_cst:
+            get = atomicLoad!(memory_order.memory_order_relaxed)(cast(A*)obj);
+            do
+            {
+                set = get | arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_seq_cst, memory_order.memory_order_seq_cst)(cast(A*)obj, &get, cast(A)set));
+            break;
+    }
+
+    return get;
+}
+
+///
+unittest
+{
+    shared(int) val = 5;
+    atomic_fetch_or_explicit_impl(&val, 3, memory_order.memory_order_seq_cst);
+    assert(atomic_load_impl(&val) == 7);
+}
+
+///
+pragma(inline, true)
+A atomic_fetch_xor_impl(A, M)(shared(A)* obj, M arg) @trusted
+{
+    assert(obj !is null);
+
+    // copied from atomicOp
+
+    A set, get = atomicLoad(cast(A*)obj);
+
+    do
+    {
+        set = get ^ arg;
+    } while (!atomicCompareExchangeWeak!(memory_order.memory_order_seq_cst, memory_order.memory_order_seq_cst)(cast(A*)obj, &get, cast(A)set));
+
+    return get;
+}
+
+///
+unittest
+{
+    shared(int) val = 5;
+    atomic_fetch_xor_impl(&val, 3);
+    assert(atomic_load_impl(&val) == 6);
+}
+
+///
+pragma(inline, true)
+A atomic_fetch_xor_explicit_impl(A, M)(shared(A)* obj, M arg, memory_order order) @trusted
+{
+    assert(obj !is null);
+
+    A set, get;
+
+    final switch(order)
+    {
+        case memory_order.memory_order_relaxed:
+            get = atomicLoad!(memory_order.memory_order_relaxed)(cast(A*)obj);
+            do
+            {
+                set = get ^ arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_relaxed, memory_order.memory_order_relaxed)(cast(A*)obj, &get, cast(A)set));
+            break;
+
+        case memory_order.memory_order_acquire:
+            get = atomicLoad!(memory_order.memory_order_acquire)(cast(A*)obj);
+            do
+            {
+                set = get ^ arg;
+            } while (!atomicCompareExchangeWeak!(memory_order.memory_order_acquire, memory_order.memory_order_acquire)(cast(A*)obj, &get, cast(A)set));
+            break;
```

Comment:
This would be better if can be made into a template or mixin.

The only cases that should really matter are:
```
if (order == memory_order.memory_order_relaxed)
{
    // Instantiate with atomicCompare!(memory_order.memory_order_relaxed)
}
else if (order != memory_order.memory_order_seq_cst)
{
    // Instantiate with atomicCompare!(memory_order.memory_order_acq_rel)
}
else
{
    // Instantiate with atomicCompare!(memory_order.memory_order_seq_cst)
}
```

Reply:
Perhaps it can be rewritten if someone cares about it. However, I'm certain that this design has the best chance to optimize in both ldc and gdc.

> the memory synchronization ordering for the load operation if the comparison fails. Cannot be [memory_order_release](https://en.cppreference.com/w/c/atomic/memory_order) or [memory_order_acq_rel](https://en.cppreference.com/w/c/atomic/memory_order) and cannot specify stronger ordering than succ

Okay I can simplify the fail down.